runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/middleware/Retry.kt

```diff
@@ -11,31 +11,67 @@ import aws.smithy.kotlin.runtime.http.HttpBody
 import aws.smithy.kotlin.runtime.http.HttpClientFeatureFactory
 import aws.smithy.kotlin.runtime.http.operation.SdkHttpOperation
 import aws.smithy.kotlin.runtime.http.operation.deepCopy
+import aws.smithy.kotlin.runtime.http.operation.getLogger
 import aws.smithy.kotlin.runtime.http.request.HttpRequestBuilder
+import aws.smithy.kotlin.runtime.logging.Logger
+import aws.smithy.kotlin.runtime.retries.RetryDirective
 import aws.smithy.kotlin.runtime.retries.RetryPolicy
 import aws.smithy.kotlin.runtime.retries.RetryStrategy
+import aws.smithy.kotlin.runtime.util.InternalApi
+
+/**
+ * Retry requests with the given strategy and policy
+ * @param strategy the [RetryStrategy] to retry failed requests with
+ * @param policy the [RetryPolicy] used to determine when to retry
+ */
+@InternalApi
+class Retry(
+    private val strategy: RetryStrategy,
+    private val policy: RetryPolicy<Any?>
+) : Feature {
 
-class RetryFeature(private val strategy: RetryStrategy, private val policy: RetryPolicy<Any?>) : Feature {
     class Config {
         var strategy: RetryStrategy? = null
         var policy: RetryPolicy<Any?>? = null
     }
 
-    companion object Feature : HttpClientFeatureFactory<Config, RetryFeature> {
-        override val key: FeatureKey<RetryFeature> = FeatureKey("RetryFeature")
+    companion object Feature : HttpClientFeatureFactory<Config, Retry> {
+        override val key: FeatureKey<Retry> = FeatureKey("Retry")
 
-        override fun create(block: Config.() -> Unit): RetryFeature {
+        override fun create(block: Config.() -> Unit): Retry {
             val config = Config().apply(block)
             val strategy = requireNotNull(config.strategy) { "strategy is required" }
             val policy = requireNotNull(config.policy) { "policy is required" }
-            return RetryFeature(strategy, policy)
+            return Retry(strategy, policy)
+        }
+    }
+
+    /**
+     * Wrapper around [policy] that logs termination decisions
+     */
+    private class PolicyLogger(
+        private val policy: RetryPolicy<Any?>,
+        private val logger: Logger,
+    ) : RetryPolicy<Any?> {
+        override fun evaluate(result: Result<Any?>): RetryDirective = policy.evaluate(result).also {
+            if (it is RetryDirective.TerminateAndFail) {
+                logger.debug { "request failed with non-retryable error" }
```

Comment:
suggestion

emit the error that was non-retryable

Reply:
It should throw with the error already right?